lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1058,6 +1058,36 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
     }
 
     pattern = newPattern;
+
+    // Use of 'as!' or 'as?' will cause errors in the type checking below.
+    // Catch that here and give a clear error message.
+    if (auto VarPat = dyn_cast<VarPattern>(pattern)) {
+      if (auto ExprPat = dyn_cast<ExprPattern>(VarPat->getSubPattern())) {
+        if (auto SubExpr = ExprPat->getSubExpr())
+          if (auto SeqExpr = dyn_cast<SequenceExpr>(SubExpr)) {
+            SourceLoc LastPunctLoc;
+            SeqExpr->forEachImmediateChildExpr([&](Expr *E) -> Expr * {
```

Comment:
I'm a little worried here about some funky sequence like `1 + 2 as? Int`. It might be safer to only do this check if the SequenceExpr has exactly two elements. (And then you don't have to worry about repeat diagnostics too, right?)